From 0f453b7495f04cb5e5613e05c0c60631bb59ec53 Mon Sep 17 00:00:00 2001
From: Vignesh Sridharan <vignsrid@qti.qualcomm.com>
Date: Fri, 8 May 2026 12:21:55 +0530
Subject: [PATCH] bwl/bpl: Write SMD policy parameters to hostapd and UCI
 config

Write SMD (Seamless Mobility Domain) policy parameters to hostapd
configuration and UCI wireless config for persistence.

- ap_wlan_hal_nl80211.cpp: Write SMD params in update_vap_credentials()
  and create_new_hostapd_conf_file(); clear params when not applicable
- bpl_cfg_wifi.cpp: Persist SMD params in bpl_cfg_set_wifi_credentials()
  for SLO path, MLD wifi-iface path, and MLD wifi-mld path

Parameters: smd_ap, smd_identifier, smd_timeout, smd_dl_data_fwd,
smd_max_peer_apmlds, smd_type

Signed-off-by: Vignesh Sridharan <vignsrid@qti.qualcomm.com>
---
 .../bwl/nl80211/ap_wlan_hal_nl80211.cpp       | 46 +++++++++++++++++++
 .../platform/bpl/cfg/uci/bpl_cfg_wifi.cpp     |  9 ++++
 2 files changed, 55 insertions(+)

diff --git a/common/beerocks/bwl/nl80211/ap_wlan_hal_nl80211.cpp b/common/beerocks/bwl/nl80211/ap_wlan_hal_nl80211.cpp
index bba4e3bdc..651edac9f 100644
--- a/common/beerocks/bwl/nl80211/ap_wlan_hal_nl80211.cpp
+++ b/common/beerocks/bwl/nl80211/ap_wlan_hal_nl80211.cpp
@@ -2021,6 +2021,34 @@ bool ap_wlan_hal_nl80211::update_vap_credentials(
                         build_multiap_assocresp_ie(hop_count));
                 }
 
+                // Write SMD policy parameters to hostapd config
+                if (bss_it.has_smd_policy) {
+                    conf.set_create_vap_value(vap_id, "smd_ap", "1");
+                    conf.set_create_vap_value(vap_id, "smd_identifier",
+                                              tlvf::mac_to_string(bss_it.smd_identifier));
+                    conf.set_create_vap_value(vap_id, "smd_timeout",
+                                              std::to_string(bss_it.smd_timeout));
+                    conf.set_create_vap_value(vap_id, "smd_dl_data_fwd",
+                                              std::to_string(bss_it.smd_dl_forwarding));
+                    conf.set_create_vap_value(vap_id, "smd_max_peer_apmlds",
+                                              std::to_string(bss_it.smd_max_prepared_ap_mlds));
+                    conf.set_create_vap_value(vap_id, "smd_type", std::to_string(bss_it.smd_type));
+                    LOG(DEBUG) << "Written SMD config to hostapd for BSS " << bss_it.bssid
+                               << ": smd_identifier=" << bss_it.smd_identifier
+                               << ", timeout=" << bss_it.smd_timeout
+                               << ", dl_data_fwd=" << int(bss_it.smd_dl_forwarding)
+                               << ", max_peer_apmlds=" << int(bss_it.smd_max_prepared_ap_mlds)
+                               << ", type=" << int(bss_it.smd_type);
+                } else {
+                    // Clear SMD parameters when not applicable
+                    conf.set_create_vap_value(vap_id, "smd_ap", "");
+                    conf.set_create_vap_value(vap_id, "smd_identifier", "");
+                    conf.set_create_vap_value(vap_id, "smd_timeout", "");
+                    conf.set_create_vap_value(vap_id, "smd_dl_data_fwd", "");
+                    conf.set_create_vap_value(vap_id, "smd_max_peer_apmlds", "");
+                    conf.set_create_vap_value(vap_id, "smd_type", "");
+                }
+
                 //required to update uci wireless
                 bss_it.multi_ap = multi_ap;
                 // oddly enough, multi_ap_backhaul_wpa_passphrase has to be
@@ -3789,6 +3817,24 @@ bool ap_wlan_hal_nl80211::create_new_hostapd_conf_file(std::string &ifname, son:
     conf.set_create_vap_value(vap_id, "rrm_neighbor_report", 1);
     conf.set_create_vap_value(vap_id, "rrm_beacon_report", 1);
 
+    // Write SMD policy parameters to hostapd config
+    if (bss_conf.has_smd_policy) {
+        conf.set_create_vap_value(vap_id, "smd_ap", "1");
+        conf.set_create_vap_value(vap_id, "smd_identifier",
+                                  tlvf::mac_to_string(bss_conf.smd_identifier));
+        conf.set_create_vap_value(vap_id, "smd_timeout", std::to_string(bss_conf.smd_timeout));
+        conf.set_create_vap_value(vap_id, "smd_dl_data_fwd",
+                                  std::to_string(bss_conf.smd_dl_forwarding));
+        conf.set_create_vap_value(vap_id, "smd_max_peer_apmlds",
+                                  std::to_string(bss_conf.smd_max_prepared_ap_mlds));
+        conf.set_create_vap_value(vap_id, "smd_type", std::to_string(bss_conf.smd_type));
+        LOG(DEBUG) << "Written SMD config to hostapd for new BSS " << bss_conf.bssid
+                   << ": smd_identifier=" << bss_conf.smd_identifier
+                   << ", timeout=" << bss_conf.smd_timeout
+                   << ", dl_data_fwd=" << int(bss_conf.smd_dl_forwarding)
+                   << ", max_peer_apmlds=" << int(bss_conf.smd_max_prepared_ap_mlds)
+                   << ", type=" << int(bss_conf.smd_type);
+    }
 
     if (!assign_auth_encr_parameters(conf,vap_id, bss_conf)) {
         LOG(ERROR) << "Failed to set the authentication/encryption parameters!";
diff --git a/framework/platform/bpl/cfg/uci/bpl_cfg_wifi.cpp b/framework/platform/bpl/cfg/uci/bpl_cfg_wifi.cpp
index 306c4bee8..030db95fd 100644
--- a/framework/platform/bpl/cfg/uci/bpl_cfg_wifi.cpp
+++ b/framework/platform/bpl/cfg/uci/bpl_cfg_wifi.cpp
@@ -712,6 +712,15 @@ bool bpl_cfg_set_wifi_credentials(const std::string &iface,
                     options["custom_alias"] = vap_type_str;
                 }
             }
+            // SMD policy parameters
+            if (config.has_smd_policy) {
+                options["smd_ap"]              = "1";
+                options["smd_identifier"]      = utils::mac_to_string(config.smd_identifier.oct);
+                options["smd_timeout"]         = std::to_string(config.smd_timeout);
+                options["smd_dl_data_fwd"]     = std::to_string(config.smd_dl_forwarding);
+                options["smd_max_peer_apmlds"] = std::to_string(config.smd_max_prepared_ap_mlds);
+                options["smd_type"]            = std::to_string(config.smd_type);
+            }
 
             return options;
         };
-- 
2.34.1

